net/src/ipc/n3h.rs

```diff
@@ -0,0 +1,294 @@
+        _ => bail!("arch {} not available", arch),
+    };
+    if arch.is_none() {
+        bail!("arch not available");
+    }
+    Ok(&arch.as_ref().unwrap())
+}
+
+/// run a command / capture the output
+fn exec_output<S1, I, S2>(cmd: S1, args: I, ignore_errors: bool) -> NetResult<String>
+where
+    S1: AsRef<std::ffi::OsStr>,
+    I: IntoIterator<Item = S2>,
+    S2: AsRef<std::ffi::OsStr>,
+{
+    let mut cmd = std::process::Command::new(cmd);
+    cmd.args(args);
+    tlog_d!("EXEC: {:?}", cmd);
+    let res = cmd.output()?;
+    if !ignore_errors && !res.status.success() {
+        bail!(
+            "bad exit {:?} {:?}",
+            res.status.code(),
+            String::from_utf8_lossy(&res.stderr)
+        );
+    }
+    Ok(String::from_utf8_lossy(&res.stdout).trim().to_string())
+}
+
+/// get the current os / arch
+fn get_os_arch() -> NetResult<(&'static str, &'static str)> {
+    if cfg!(target_os = "linux") && cfg!(target_arch = "x86_64") {
+        Ok(("linux", "x64"))
+    } else if cfg!(target_os = "linux") && cfg!(target_arch = "aarch64") {
+        Ok(("linux", "aarch64"))
+    } else if cfg!(target_os = "windows") && cfg!(target_arch = "x86_64") {
+        Ok(("win", "x64"))
+    } else if cfg!(target_os = "macos") && cfg!(target_arch = "x86_64") {
+        Ok(("mac", "x64"))
+    } else {
+        bail!("no prebuilt n3h for current os/arch - TODO check for node/npm - dl release zip");
+    }
+}
+
+/// extract a dmg archive
+fn extract_dmg(file: &std::ffi::OsStr, dest: &std::path::PathBuf) -> NetResult<std::path::PathBuf> {
+    let mut dest = dest.clone();
+    dest.push("n3h.app");
+    if !dest.exists() {
+        tlog_d!(
+            "{}",
+            exec_output(
+                "hdiutil",
+                vec![
+                    "attach",
+                    "-mountpoint",
+                    "./dmg-mount",
+                    &file.to_string_lossy()
+                ],
+                false
+            )?
+        );
+        exec_output(
+            "cp",
+            vec!["-a", "./dmg-mount/n3h.app", &dest.to_string_lossy()],
+            true,
+        )?;
+        tlog_d!(
+            "{}",
+            exec_output("hdiutil", vec!["detach", "./dmg-mount"], false)?
+        );
+    }
+    dest.push("Contents");
+    dest.push("MacOS");
+    dest.push("n3h");
+    Ok(dest)
+}
+
+/// hash a file && compare to expected hash
+fn check_hash(file: &std::ffi::OsStr, sha256: &str) -> bool {
+    let mut file = match std::fs::File::open(file) {
+        Err(_) => return false,
+        Ok(v) => v,
+    };
+
+    let mut hash = sha2::Sha256::new();
+
+    if std::io::copy(&mut file, &mut hash).is_err() {
+        return false;
+    }
+
+    let hash = format!("{:x}", hash.result());
+
+    if &hash != sha256 {
+        tlog_e!("bad hash, expected {}, got {}", sha256, &hash);
+    }
+
+    &hash == sha256
+}
+
+/// hash a file, if it doesn't exist or fails, download && hash
```

Comment:
`Download a missing file from disk. Check is done by hashing the file's content.`

Reply:
updating comment :+1: